### ISSUE TITLE
Friendlier API for A<T>.That.IsSameSequenceAs

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#1194]
From comment: https://github.com/FakeItEasy/FakeItEasy/issues/1193#issuecomment-325925228

Currently `IsSameSequenceAs` takes an non-generic `IEnumerable`. This is less than ideal, because:
- it doesn't ensure type safety, e.g. you can do `IsSameSequenceAs(new[]{ "hello", "world" })` when the parameter is actually a list of ints.
- it's cumbersome to type; in the typical case you'd just want to write `IsSameSequenceAs(1, 2, 3)` rather than `IsSameSequenceAs(new[]{1, 2, 3})`

Proposal:

- add an overload that works on `IEnumerable<T>`
- add overloads which accept the expected elements as a `params` array

New overloads:

```csharp
public static T IsSameSequenceAs<T, TItem>(
    this IArgumentConstraintManager<T> manager,
    IEnumerable<TItem> values) where T : IEnumerable<TItem>;
public static T IsSameSequenceAs<T, TItem>(
    this IArgumentConstraintManager<T> manager,
    params TItem[] values) where T : IEnumerable<TItem>;
public static T IsSameSequenceAs<T>(
    this IArgumentConstraintManager<T> manager,
    params object[] values) where T : IEnumerable;
```

**Comment by blairconrad** [comment]
Love it, @thomaslevesque. Seems well-enough defined that we could up-for-grabs it?

**Comment by thomaslevesque** [comment]
Definitely!

**Comment by thomaslevesque** [comment]
Adding the new overloads as an unexpected benefit... Overload resolution picks the `IsSameSequenceAs<T, TItem>` overloads over the `IsSameSequenceAs<T>` overloads, which means that these statements won't compile:

    A.CallTo(() => foo.Bar(A<IEnumerable<int>>.That.IsSameSequenceAs("hello", "world"))).Returns(42);
    A.CallTo(() => foo.Bar(A<IEnumerable<int>>.That.IsSameSequenceAs(new[]{"hello", "world"}))).Returns(42);

I was afraid it would pick the overloads with non-generic `IEnumerable` in those cases, but it doesn't. I think this is the ideal behavior.

The bad news is that it's a breaking change: the second line in the snippet above compiles with the current version, and with this change, it won't. But I think it's a "good" breaking change, because such code is obviously incorrect and would have no useful effect.

**Comment by thomaslevesque** [comment]
Ah. More bad news... These seemingly valid statements won't compile with the proposed change:

    A.CallTo(() => foo.Bar(A<IEnumerable>.That.IsSameSequenceAs("hello", "world"))).Returns(42);
    A.CallTo(() => foo.Bar(A<IEnumerable>.That.IsSameSequenceAs(new[]{"hello", "world"}))).Returns(42);

Again, overload resolution picks `IsSameSequenceAs<T, TItem>`, and the non-generic `IEnumerable` doesn't satisfy the constraint (`where T : IEnumerable<TItem>`).

This is a more serious issue IMO. This makes the non-generic overloads very awkward to use, because you have to specify the arguments like this:

    A.CallTo(() => foo.Bar(A<IEnumerable>.That.IsSameSequenceAs((object)"hello", "world"))).Returns(42);
    A.CallTo(() => foo.Bar(A<IEnumerable>.That.IsSameSequenceAs(new object[]{"hello", "world"}))).Returns(42);

Admittedly, it's pretty rare nowadays to use `IEnumerable` instead of `IEnumerable<T>`, but it's still a legitimate use case. The change would make things better for `IEnumerable<T>`, but worse for `IEnumerable`.

**Comment by thomaslevesque** [comment]
Removing `up-for-grabs`, since it needs more discussion.

**Comment by blairconrad** [comment]
> thomaslevesque (Thomas Levesque) added this to the vNext milestone 2 days ago

is there a reason we wouldn't want to release without this issue?

**Comment by thomaslevesque** [comment]
> is there a reason we wouldn't want to release without this issue?

No, I just thought that since it's quick enough to implement, we might as well do it in the next release.

But as I said above, the overloads that work on `IEnumerable<T>` make the ones that work on `IEnumerable` difficult to use, so maybe we shouldn't implement them... We can still add a `params object[]` overload, though.

**Comment by thomaslevesque** [comment]
We could add the generic overloads with a different name, but I can't think of anything as good as `IsSameSequenceAs`, and it would be confusing...

**Comment by blairconrad** [comment]
> We can still add a `params object[]` overload, though.

I like this idea.

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 4.1.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/4.1.0).
